File: sdk/python/core/tests/test_utils.py
#  ----------------------------------------------------------------
# Copyright 2016 Cisco Systems
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------

"""test_utils.py

Utility function for test cases.
"""
import re
import unittest
import logging

from argparse import ArgumentParser

import sys
if sys.version_info > (3,):
    from urllib.parse import urlparse
else:
    from urlparse import urlparse

from ydk.entity_utils import get_data_node_from_entity
from ydk.errors       import YCoreError


def assert_with_error(pattern, ErrorClass):
    def assert_with_pattern(func):
        def helper(self, *args, **kwargs):
            try:
                func(self)
            except ErrorClass as error:
                if hasattr(error, 'message'):
                    res = re.match(pattern, error.message.strip())
                elif hasattr(error, 'args'):
                    res = pattern in error.args[0]
                self.assertTrue(res)
        return helper
    return assert_with_pattern


class ParametrizedTestCase(unittest.TestCase):
    """ TestCase classes that want to be parametrized should
        inherit from this class.
    """
    def __init__(self, method_name='runTest'):
        if sys.version_info > (3,):
            super().__init__(method_name)
        else:
            super(ParametrizedTestCase, self).__init__(method_name)

    @staticmethod
    def parametrize(testcase_klass, device, non_demand, common_cache, timeout):
        """ Create a suite containing all tests taken from the given
            subclass, passing them the parameter 'param'.
        """
        testloader = unittest.TestLoader()
        testcase_klass.hostname = device.hostname
        testcase_klass.username = device.username
        testcase_klass.password = device.password
        testcase_klass.port = device.port
        testcase_klass.protocol = device.scheme
        testcase_klass.on_demand = not non_demand
        testcase_klass.common_cache = common_cache
        testcase_klass.timeout = timeout
        testnames = testloader.getTestCaseNames(testcase_klass)
        suite = unittest.TestSuite()
        for name in testnames:
            suite.addTest(testcase_klass(name))
        return suite


def get_device_info():
    parser = ArgumentParser()
    parser.add_argument("-v", "--verbose", help="print debugging messages",
                        action="store_true")
    parser.add_argument("--non-demand", help="disable on demand model downloading",
                        dest="non_demand", action="store_true")
    parser.add_argument("--common-cache", help="use common cache directory",
                        dest="common_cache", action="store_true")
    parser.add_argument("device", nargs='?',
                        help="NETCONF device (ssh://user:password@host:port)")
    parser.add_argument("--timeout", help="timeout in microseconds, -1 for infinite timeout, 0 for non-blocking",
                        dest="timeout", action='store', type=int, default=-1)

    args = parser.parse_args()
    if not args.device:
        args.device = "ssh://admin:admin@127.0.0.1:12022"
    device = urlparse(args.device)
    return device, args.non_demand, args.common_cache, args.timeout


def datanode_to_str(dn, indent = ''):
    try:
        s = dn.get_schema_node().get_statement()
        if s.keyword == "leaf" or s.keyword == "leaf-list" or s.keyword == "anyxml":
            out = indent + "<" + s.arg + ">" + dn.get_value() + "</" + s.arg + ">\n"
        else:
            out = indent + "<" + s.arg + ">\n"
            child_indent = indent + "  "
            for child in dn.get_children():
                out += datanode_to_str(child, child_indent)
            out += indent + "</" + s.arg + ">\n"
        return out
    except YCoreError as ex:
        print(ex.message)


def print_data_node(dn):
    try:
        print("\n=====>  Printing DataNode: '{}'".format(dn.get_path()))
        print(datanode_to_str(dn))
    except YCoreError as ex:
        print(ex.message)


def print_entity(entity, root_schema):
    dn = get_data_node_from_entity( entity, root_schema);
    print_data_node(dn)


def enable_logging(level):
    log = logging.getLogger('ydk')
    log.setLevel(level)
    handler = logging.StreamHandler()
    formatter = logging.Formatter(("%(asctime)s - %(name)s - %(levelname)s - %(message)s"))
    handler.setFormatter(formatter)
    log.addHandler(handler)

File: sdk/cpp/tests/test_utils.cpp
/// YANG Development Kit
// Copyright 2016 Cisco Systems. All rights reserved
//
////////////////////////////////////////////////////////////////
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
//////////////////////////////////////////////////////////////////
#include <sstream>

#include <ydk/entity_data_node_walker.hpp>
#include <ydk/filters.hpp>

#include "test_utils.hpp"

using namespace std;

static std::string yfilter2str(ydk::path::DataNode* dn)
{
    auto filter = ydk::get_data_node_yfilter(dn);
    if (filter == ydk::YFilter::not_set)
        return "";
    std::string filter_str = " yfilter=";
    filter_str += to_string(filter);
    return filter_str;
}

std::string print_tree(ydk::path::DataNode* dn, const std::string& indent)
{
    std::ostringstream buffer;
    try {
        auto filter = yfilter2str(dn);
        ydk::path::Statement s = dn->get_schema_node().get_statement();
        if(s.keyword == "leaf" || s.keyword == "leaf-list" || s.keyword == "anyxml") {
            auto val = dn->get_value();
            buffer << indent << "<" << s.arg << filter << ">" << val << "</" << s.arg << ">" << std::endl;
        }
         else {
            std::string child_indent{indent};
            child_indent+="  ";
            buffer << indent << "<" << s.arg << filter << ">" << std::endl;
            for(auto c : dn->get_children())
                buffer << print_tree(c.get(), child_indent);
            buffer << indent << "</" << s.arg << ">" << std::endl;
        }
    }
    catch (ydk::path::YCoreError &ex) {
        cout << ex.what() << endl;
    }
    return buffer.str();
}

void print_data_node(shared_ptr<ydk::path::DataNode> dn)
{
    cout << data_node_to_xml(dn) << endl;
}

std::string data_node_to_xml(shared_ptr<ydk::path::DataNode> dn)
{
    std::ostringstream buffer;
    try {
        buffer << "\n=====>  Printing DataNode: '" << dn->get_path() << "'" << endl;
        buffer << print_tree(dn.get(), " ");
    }
    catch (ydk::path::YCoreError &ex) {
        std::cout << ex.what() << std::endl;
    }
    return buffer.str();
}

std::string data_node_to_xml(ydk::path::DataNode & dn)
{
    std::ostringstream buffer;
    try {
        buffer << "\n=====>  Printing DataNode: '" << dn.get_path() << "'" << endl;
        buffer << print_tree(&dn, " ");
    }
    catch (ydk::path::YCoreError &ex) {
        std::cout << ex.what() << std::endl;
    }
    return buffer.str();
}

void print_entity(shared_ptr<ydk::Entity> entity, ydk::path::RootSchemaNode& root)
{
    ydk::path::DataNode& dn = get_data_node_from_entity( *entity, root);
    ydk::path::Statement s = dn.get_schema_node().get_statement();
    cout << "\n=====>  Printing DataNode: '" << s.arg << "'" << endl;
    cout << print_tree( &dn, " ");
}

string entity2string(shared_ptr<ydk::Entity> entity, ydk::path::RootSchemaNode& root)
{
    ydk::path::DataNode& dn = get_data_node_from_entity( *entity, root);
    ydk::path::Statement s = dn.get_schema_node().get_statement();
    return print_tree( &dn, "");
}
